rs/execution_environment/src/canister_manager/tests.rs

```diff
@@ -4406,19 +4404,37 @@ fn cycles_correct_if_install_fails_at_validation() {
     let initial_cycles = Cycles::new(1_000_000_000_000_000);
     let id = test.create_canister(initial_cycles);
 
-    test.install_canister_with_allocation(id, wasm, Some(100), None)
-        .unwrap_err();
+    test.install_canister(id, wasm.clone()).unwrap();
     assert_eq!(
         test.canister_state(id).system_state.balance(),
         initial_cycles - test.canister_execution_cost(id),
     );
-    assert_eq!(
+    let execution_cost_of_successful_installation = test.cycles_account_manager().execution_cost(
+        wasm_compilation_cost(&wasm) + NumInstructions::from(5 * *DROP_MEMORY_GROW_CONST_COST),
+        test.subnet_size(),
+        test.canister_wasm_execution_mode(id),
+    );
+    assert_delta!(
         test.canister_execution_cost(id),
-        test.cycles_account_manager().execution_cost(
-            NumInstructions::from(0),
-            test.subnet_size(),
-            test.canister_wasm_execution_mode(id),
-        )
+        execution_cost_of_successful_installation,
+        Cycles::new(10)
+    );
+
+    // Attempt to install the same canister again, should fail because it's non-empty.
```

Comment:
How about failing the validation due to invalid controllers? Then we wouldn't need a pair of installations.

Reply:
That was my first attempt :) But since we send the message as an ingress, it is rejected way before it reaches the point of validations we care about in this test (`CanisterManager::should_accept_ingress` throws it out already). Then nothing is really charged (as expected for such a case) and the test would be trivially succeeding.

Reply:
I see your point (and don't see a better way).